probe-rs/src/probe/arm_debug_interface.rs

```diff
@@ -508,6 +494,93 @@ fn perform_transfers<P: DebugProbe + RawProtocolIo + JTAGAccess>(
     Ok(())
 }
 
+/// Perform a batch of raw transfers, retrying on WAIT responses.
+///
+/// Other than that, the transfers are sent as-is. You might want to use `perform_transfers` instead, which
+/// does correction for delayed FAULT responses and other helpful stuff.
+fn perform_raw_transfers_retry<P: DebugProbe + RawProtocolIo + JTAGAccess>(
+    probe: &mut P,
+    transfers: &mut [DapTransfer],
+) -> Result<(), ArmError> {
+    let mut successful_transfers = 0;
+    let mut idle_cycles = std::cmp::max(1, probe.swd_settings().num_idle_cycles_between_writes);
+
+    'transfer: for _ in 0..probe.swd_settings().num_retries_after_wait {
+        let chunk = &mut transfers[successful_transfers..];
+        assert!(!chunk.is_empty());
+
+        // Set idle cycles
+        for transfer in &mut chunk[..] {
+            if transfer.is_write() {
+                // TODO this doesn't quite match the old behavior, is that OK?
+                transfer.idle_cycles_after = transfer.idle_cycles_after.max(idle_cycles);
+            }
+        }
+
+        perform_raw_transfers(probe, chunk)?;
+
+        for result in chunk.iter() {
+            match result.status {
+                TransferStatus::Ok => {
+                    successful_transfers += 1;
+                }
+                TransferStatus::Failed(DapError::WaitResponse) => {
+                    tracing::debug!("got WAIT on transfer {}, retrying...", successful_transfers);
+
+                    clear_overrun(probe)?;
+
+                    idle_cycles = std::cmp::min(
+                        probe.swd_settings().max_retry_idle_cycles_after_wait,
+                        idle_cycles * 2,
+                    );
+
+                    continue 'transfer;
+                }
+                _ => break, // on any other error, we're done.
```

Comment:
@Dirbaio do you remember if this was supposed to break out of the `'transfer` loop, or retry ad infinitum (1000 times)? Looks like a bug and I'm hitting it when trying to talk to a RISC-V chip while the code thinks it's ARM (during detection).

Reply:
hmmm yeah it looks like it should be a return, yes.